<a href="https://colab.research.google.com/github/Soichiro-Gardinner/Neural_Networks/blob/main/Neural_NetwoerkKaggle.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Hausing Dataset**
- **By:** Oscar Castanaza

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("/content/train.csv")
df.head(11)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0    1          60       RL         65.0     8450   Pave   NaN      Reg   
1    2          20       RL         80.0     9600   Pave   NaN      Reg   
2    3          60       RL         68.0    11250   Pave   NaN      IR1   
3    4          70       RL         60.0     9550   Pave   NaN      IR1   
4    5          60       RL         84.0    14260   Pave   NaN      IR1   
5    6          50       RL         85.0    14115   Pave   NaN      IR1   
6    7          20       RL         75.0    10084   Pave   NaN      Reg   
7    8          60       RL          NaN    10382   Pave   NaN      IR1   
8    9          50       RM         51.0     6120   Pave   NaN      Reg   
9   10         190       RL         50.0     7420   Pave   NaN      Reg   
10  11          20       RL         70.0    11200   Pave   NaN      Reg   

   LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
5          Lvl    AllPub  ...        0    NaN  MnPrv        Shed     700   
6          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
7          Lvl    AllPub  ...        0    NaN    NaN        Shed     350   
8          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
9          Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
10         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

   MoSold YrSold  SaleType  SaleCondition  SalePrice  
0       2   2008        WD         Normal     208500  
1       5   2007        WD         Normal     181500  
2       9   2008        WD         Normal     223500  
3       2   2006        WD        Abnorml     140000  
4      12   2008        WD         Normal     250000  
5      10   2009        WD         Normal     143000  
6       8   2007        WD         Normal     307000  
7      11   2009        WD         Normal     200000  
8       4   2008        WD        Abnorml     129900  
9       1   2008        WD         Normal     118000  
10      2   2008        WD         Normal     129500  

[11 rows x 81 columns]

In [ ]:
df.isna().sum()

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [ ]:
df.shape

(1460, 81)

# **Split and Impute**

- Impute

In [ ]:
from sklearn.impute import SimpleImputer

# Separate the features (X) and target variable (y)
X = df.drop('SalePrice', axis=1)
y = df['SalePrice']

# Identify the numeric and categorical columns
numeric_cols = X.select_dtypes(include='number').columns
categorical_cols = X.select_dtypes(include='object').columns

# Impute missing values in numeric columns with mean imputation
numeric_imputer = SimpleImputer(strategy='mean')
X[numeric_cols] = numeric_imputer.fit_transform(X[numeric_cols])

# Impute missing values in categorical columns with mode imputation
categorical_imputer = SimpleImputer(strategy='most_frequent')
X[categorical_cols] = categorical_imputer.fit_transform(X[categorical_cols])

# Split the imputed data into train and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## **Scale, OH encoder, ColumnTransformer and Pipeline.**

In [ ]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Preprocessing pipeline for numeric features
numeric_features = ['LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
                    'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF',
                    '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
                    'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt',
                    'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch',
                    'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold']

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

# Preprocessing pipeline for categorical features
categorical_features = ['MSSubClass', 'MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
                        'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
                        'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
                        'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure',
                        'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical',
                        'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
                        'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType',
                        'SaleCondition']

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('encoder', OneHotEncoder(handle_unknown='ignore'))
])

# Create a preprocessor to apply the transformations
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

# Apply the preprocessing to the training and test data
X_train_preprocessed = preprocessor.fit_transform(X_train)
X_test_preprocessed = preprocessor.transform(X_test)

# **Models**

- Model 1

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras import regularizers

# Define the architecture of the deep learning model
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(X_train_preprocessed.shape[1],)))
model.add(Dropout(0.2))
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(1))

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
history = model.fit(X_train_preprocessed.toarray(), y_train, epochs=100, batch_size=32, validation_data=(X_test_preprocessed.toarray(), y_test))


# Evaluate the model
loss = model.evaluate(X_test_preprocessed.toarray(), y_test.values)
print('\nMean Squared Error:', loss)

Epoch 1/100
37/37 [==============================] - 2s 15ms/step - loss: 38883160064.0000 - val_loss: 39645880320.0000
Epoch 2/100
37/37 [==============================] - 0s 9ms/step - loss: 38858674176.0000 - val_loss: 39594762240.0000
Epoch 3/100
37/37 [==============================] - 0s 12ms/step - loss: 38757244928.0000 - val_loss: 39428358144.0000
Epoch 4/100
37/37 [==============================] - 0s 9ms/step - loss: 38496329728.0000 - val_loss: 39054110720.0000
Epoch 5/100
37/37 [==============================] - 0s 7ms/step - loss: 37974491136.0000 - val_loss: 38376656896.0000
Epoch 6/100
37/37 [==============================] - 0s 9ms/step - loss: 37092888576.0000 - val_loss: 37296721920.0000
Epoch 7/100
37/37 [==============================] - 0s 5ms/step - loss: 35791855616.0000 - val_loss: 35763359744.0000
Epoch 8/100
37/37 [==============================] - 0s 5ms/step - loss: 33983461376.0000 - val_loss: 33747650560.0000
Epoch 9/100
37/37 [===========================

- Model 2

In [ ]:
# Define the architecture of the deep learning model
model2 = Sequential()
model2.add(Dense(64, activation='relu', input_shape=(X_train_preprocessed.shape[1],), kernel_regularizer=regularizers.l2(0.001)))
model2.add(Dense(64, activation='relu', kernel_regularizer=regularizers.l2(0.001)))
model2.add(Dense(1))

# Compile the model
model2.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
history2 = model2.fit(X_train_preprocessed.toarray(), y_train.values, epochs=100, batch_size=32, validation_data=(X_test_preprocessed.toarray(), y_test.values))

# Evaluate the model
loss2 = model2.evaluate(X_test_preprocessed.toarray(), y_test.values)
print('Model 2 - Mean Squared Error:', loss2)


Epoch 1/100
37/37 [==============================] - 1s 7ms/step - loss: 38883889152.0000 - val_loss: 39648301056.0000
Epoch 2/100
37/37 [==============================] - 0s 4ms/step - loss: 38866239488.0000 - val_loss: 39610933248.0000
Epoch 3/100
37/37 [==============================] - 0s 3ms/step - loss: 38792368128.0000 - val_loss: 39486771200.0000
Epoch 4/100
37/37 [==============================] - 0s 3ms/step - loss: 38593736704.0000 - val_loss: 39204347904.0000
Epoch 5/100
37/37 [==============================] - 0s 3ms/step - loss: 38195728384.0000 - val_loss: 38679236608.0000
Epoch 6/100
37/37 [==============================] - 0s 3ms/step - loss: 37509959680.0000 - val_loss: 37835579392.0000
Epoch 7/100
37/37 [==============================] - 0s 4ms/step - loss: 36471656448.0000 - val_loss: 36615856128.0000
Epoch 8/100
37/37 [==============================] - 0s 3ms/step - loss: 35035824128.0000 - val_loss: 34981924864.0000
Epoch 9/100
37/37 [=============================

- **Model 3:** Adding L2 Regularization (Ridge Regression)

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras import regularizers


# Define the architecture of the deep learning model
model3 = Sequential()
model3.add(Dense(64, activation='relu', input_shape=(X_train_preprocessed.shape[1],), kernel_regularizer=regularizers.l2(0.001)))
model3.add(Dense(64, activation='relu', kernel_regularizer=regularizers.l2(0.001)))
model3.add(Dense(1))

# Compile the model
model3.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
history3 = model3.fit(X_train_preprocessed.toarray(), y_train.values, epochs=100, batch_size=32, validation_data=(X_test_preprocessed.toarray(), y_test.values))

# Evaluate the model
loss3 = model3.evaluate(X_test_preprocessed.toarray(), y_test.values)
print('Model 3 - Mean Squared Error:', loss3)

Epoch 1/100
37/37 [==============================] - 1s 13ms/step - loss: 38882889728.0000 - val_loss: 39645114368.0000
Epoch 2/100
37/37 [==============================] - 0s 5ms/step - loss: 38856429568.0000 - val_loss: 39591661568.0000
Epoch 3/100
37/37 [==============================] - 0s 5ms/step - loss: 38756343808.0000 - val_loss: 39429029888.0000
Epoch 4/100
37/37 [==============================] - 0s 4ms/step - loss: 38503624704.0000 - val_loss: 39074435072.0000
Epoch 5/100
37/37 [==============================] - 0s 5ms/step - loss: 38010658816.0000 - val_loss: 38437052416.0000
Epoch 6/100
37/37 [==============================] - 0s 5ms/step - loss: 37197533184.0000 - val_loss: 37443158016.0000
Epoch 7/100
37/37 [==============================] - 0s 4ms/step - loss: 35991040000.0000 - val_loss: 36040368128.0000
Epoch 8/100
37/37 [==============================] - 0s 5ms/step - loss: 34347089920.0000 - val_loss: 34182881280.0000
Epoch 9/100
37/37 [============================

# **Select the Best Model:**

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

print('Model 1 - Mean Squared Error:', loss)
print('Model 2 - Mean Squared Error:', loss2)
print('Model 3 - Mean Squared Error:', loss3)

Model 1 - Mean Squared Error: 1201398272.0
Model 2 - Mean Squared Error: 1170308992.0
Model 3 - Mean Squared Error: 1162292864.0


- Model 1:

In [ ]:
# Predict on the test set
y_pred_1 = model.predict(X_test_preprocessed)

# Calculate metrics
mse_1 = mean_squared_error(y_test, y_pred_1)
mae_1 = mean_absolute_error(y_test, y_pred_1)
r2_1 = r2_score(y_test, y_pred_1)

# Print metrics
print("Model 1 Metrics:")
print("Mean Squared Error:", mse_1)
print("Mean Absolute Error:", mae_1)
print("R-squared Score:", r2_1)

10/10 [==============================] - 0s 3ms/step
Model 1 Metrics:
Mean Squared Error: 1201398183.798757
Mean Absolute Error: 19682.55070098459
R-squared Score: 0.8433706531707814


- Model 2:

In [ ]:
# Predict on the test set
y_pred_2 = model2.predict(X_test_preprocessed)

# Calculate metrics
mse_2 = mean_squared_error(y_test, y_pred_2)
mae_2 = mean_absolute_error(y_test, y_pred_2)
r2_2 = r2_score(y_test, y_pred_2)

# Print metrics
print("Model 2 Metrics:")
print("Mean Squared Error:", mse_2)
print("Mean Absolute Error:", mae_2)
print("R-squared Score:", r2_2)


10/10 [==============================] - 0s 2ms/step
Model 2 Metrics:
Mean Squared Error: 1170309099.3228042
Mean Absolute Error: 19644.836191673803
R-squared Score: 0.8474238164439186


- Model 3:

In [ ]:
# Predict on the test set
y_pred_3 = model3.predict(X_test_preprocessed)

# Calculate metrics
mse_3 = mean_squared_error(y_test, y_pred_3)
mae_3 = mean_absolute_error(y_test, y_pred_3)
r2_3 = r2_score(y_test, y_pred_3)

# Print metrics
print("Model 3 Metrics:")
print("Mean Squared Error:", mse_3)
print("Mean Absolute Error:", mae_3)
print("R-squared Score:", r2_3)

10/10 [==============================] - 0s 4ms/step
Model 3 Metrics:
Mean Squared Error: 1162292830.3642085
Mean Absolute Error: 19532.43358037243
R-squared Score: 0.8484689178831618


# **Selected**

Based on the metrics, Model 3 has the lowest Mean Squared Error (MSE) among the three models:

- Model 1: MSE = 1,201,398,183.80
- Model 2: MSE = 1,170,309,099.32
- Model 3: MSE = 1,162,292,830.36

Lower MSE values indicate better model performance because it means the model's predictions are closer to the actual target values. In this case, Model 3 has the lowest MSE, suggesting that it has better predictive accuracy compared to the other two models.



#### **Additionally**, the R-squared score is a measure of how well the model fits the data. Model 3 has the highest R-squared score of 0.848, which indicates that approximately 84.8% of the variance in the target variable can be explained by the model's predictions. Higher R-squared scores generally indicate a better fit to the data.

## **Therefore, based on the lower MSE and higher R-squared score, Model 3 appears to be the best model among the three.**